Removed unnecessary check in move_to_next_position for CombinationsWithReplacement

mfinnessey opened the pull request:
A small performance improvement opportunity I noticed :)

When iterating over the `positions` vector to find the position to start writing the `next_position` from, the current implementation of `move_to_next_position` for `CombinationsWithReplacement` checks that the candidate position is either 1) the last position in the vector or 2) less than the following position. However, `move_to_next_position` only generates non-decreasing `positions` vectors (and this check does not contribute to this). Therefore, this check will always pass and can thus safely be removed.

All tests still pass when running `cargo test`.

mfinnessey committed:
Removed unnecessary check in move_to_next_position for CombinationsWithReplacement

When iterating over the `positions` vector to find the position to start
writing the `next_position` from, the current implementation of
`move_to_next_position` for `CombinationsWithReplacement` checks that the
candidate position is either 1) the last position in the vector or 2)
less than the following position. However, `move_to_next_position` only
generates non-decreasing `positions` vectors (and this check does not
contribute to this). Therefore, this check will always pass and can
thus safely be removed.

All tests still pass when running `cargo test`.

olivercalder commented on src/combinations.rs:
```diff
@@ -284,13 +284,11 @@ impl<T: Ord + Clone> CombinationsWithReplacement<T> {
+            for i in index..length {
+                *self.positions.get_mut(i).unwrap() = next_position;
             }
+            return true;
```
Without the `return true` being conditional, the only `continue` in the loop occurs if `cur_position >= self.elements.len() - 1`, otherwise we'll always return `true`. Therefore, I don't see any way to ever advance any position index earlier than `self.elements.len() - 2`. And that seems wrong...

olivercalder replied:
Hmm, thinking some more, the idea here is that we only want to advance a position if all later positions values are already at the final index, otherwise we'd advance the latest of those that isn't yet at the final value.

So I think you're right, though I think a comment in the code explaining this subtlety would be helpful.

olivercalder commented on src/combinations.rs:
```diff
@@ -284,13 +284,11 @@ impl<T: Ord + Clone> CombinationsWithReplacement<T> {
             if cur_position >= self.elements.len() - 1 {
                 continue;
             }
```
```suggestion
            if cur_position >= self.elements.len() - 1 {
                // We only want to advance an earlier position if all later positions are already
                // at the final element. Otherwise, we'd advance the latest position which is not
                // yet at the final element.
                continue;
            }
```

olivercalder commented on src/combinations.rs:
```diff
@@ -284,13 +284,11 @@ impl<T: Ord + Clone> CombinationsWithReplacement<T> {
-                    *self.positions.get_mut(i).unwrap() = next_position;
-                }
-                return true;
+            let next_position = cur_position + 1;
```
```suggestion
            let next_position = cur_position + 1;
            // We know that `cur_position` is not at the final element, and any later positions
            // must be at the final element, so we can unconditionally set the current and any
            // subsequent positions to `next_position`.
```

olivercalder reviewed (commented):
Sorry for my delay on reviewing this.

I think you're right, this is correct, though it took enough thinking on my part to convince myself of that fact that I think some comments are in order to explain why it's safe to skip the check. I've left some suggestions, please let me know what you think of them, if you have other ideas.

mfinnessey committed:
Apply suggestions from code review

Add comments to clarify why we can safely remove the second check on cur_position

Co-authored-by: Oliver Calder <oliver@calder.dev>

mfinnessey commented:
No worries on the delay :) I agree that the justification for removing the second check on `cur_position` is non-trivial, so I adopted your suggested comments to clarify the change.

olivercalder commented:
Thanks! With that, I'll merge the PR and cut a new release. Thanks for the contribution!
